### PR TITLE
Subsystem: support optional ports via configs

### PR DESCRIPTION
**Pull request by hcook**
This PR adds support for specifying that individual ports are optional via parameters stored in the `Config` classes. This should allow their removal without modification of any `trait` extensions. I also:

- updated the SimAXIMem API to be a better example of using Edges
- cleaned up the various `Port` traits
- added helper methods for tying off various diplomatic bundles

(/cc @ppoocde  @mayoru)

### CODE REVIEW

File: src/main/scala/amba/ahb/Bundles.scala [hunk 1]
```diff
@@ -27,6 +27,27 @@ class AHBBundle(params: AHBBundleParameters) extends AHBBundleBase(params)
...
+        hwrite    := Bool(false)
+        haddr     := UInt(0)
+        hsize     := UInt(0)
+        hburst    := UInt(0)
+        hprot     := UInt(0)
+        hwdata    := UInt(0)
+      case _ =>
+    }
+  }
 }
```
Comment (terpstra) [inline src/main/scala/amba/ahb/Bundles.scala]: Could we use the named constants here?
htrans := TRANS_IDLE
hburst := BURST_SINGLE
hprot := PROT_DEFAULT
hresp := RESP_OKAY

File: src/main/scala/amba/axi4/Bundles.scala [hunk 2]
```diff
@@ -70,6 +70,24 @@ class AXI4Bundle(params: AXI4BundleParameters) extends AXI4BundleBase(params)
...
+      case OUTPUT =>
+        ar.valid := Bool(false)
+        aw.valid := Bool(false)
+        w.valid  := Bool(false)
+        r.ready  := Bool(true)
+        b.ready  := Bool(true)
+      case _ =>
+    }
+  }
 }
```
Comment (terpstra) [inline src/main/scala/amba/axi4/Bundles.scala]: Here you've taken the opposite policy. You allowed requests to go through (ready=true) even though you never respond AR=>R, AW=>B. This seems the opposite of what you might want. At least AHB/APB actually make progress when you say 'ready true'; they are the zero device. Here, AXI is the request black-hold device.

File: src/main/scala/tilelink/Bundles.scala [hunk 3]
```diff
@@ -191,6 +191,24 @@ class TLBundle(params: TLBundleParameters) extends TLBundleBase(params)
...
+        d.valid := Bool(false)
+      case OUTPUT =>
+        a.valid := Bool(false)
+        c.valid := Bool(false)
+        e.valid := Bool(false)
+        b.ready := Bool(true)
+        d.ready := Bool(true)
+      case _ =>
+    }
+  }
```
Comment (terpstra) [inline src/main/scala/tilelink/Bundles.scala]: Same as with the AXI ready.

File: src/main/scala/amba/ahb/Bundles.scala (hunk 1 above)
Comment (hcook) [reply src/main/scala/amba/ahb/Bundles.scala]: I'll use the constants. I think the blocking-forever policy is the right one.

File: src/main/scala/amba/axi4/Bundles.scala (hunk 2 above)
Comment (hcook) [reply src/main/scala/amba/axi4/Bundles.scala]: I think the blocking-forever policy is the right one. I'll make them consistent.

**Pull request merged**